### PR TITLE
[SYCL-MLIR] SYCL 'buffer' analysis

### PR DESCRIPTION
**Pull request by sommerlukas**
MLIR analysis to infer properties of interest about instances of `sycl::buffer` from their construction in host code.

### CODE REVIEW

**Comment by sommerlukas** [comment]
This PR depends on https://github.com/intel/llvm/pull/10298 and is therefore marked as draft.

**Comment by sommerlukas** [comment]
I'm aware that there's some code overlap with the ID and range analysis right now. I'm currently working on a third analysis in this series, for `sycl::accessor`. My plan is to refactor commonalities once that third analysis is complete and it is clear which parts can actually be shared, e.g., in a common base class for these three analyses.

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                    llvm::ArrayRef<size_t> constSubBufferOffset);
+
+  /// Returns true if this buffer is always constructed with the same range as
+  /// size.
+  bool hasConstantSize() const { return !constantSize.empty(); }
+
+  /// Returns the constant values for the size this buffer is constructed with.
+  const llvm::SmallVector<size_t, 3> &getConstantSize() const {
+    return constantSize;
+  }
```
Comment (victor-eds) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h:50]:
```suggestion
  /// Returns the constant values for the size this buffer is constructed with.
  llvm::ArrayRef<size_t> getConstantSize() const {
    return constantSize;
  }
```

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    assert(hasKnownBaseBuffer() && "Base buffer unknown or not unique");
+    return baseBuffer;
+  }
+
+  /// Returns true if the the base buffer of this sub-buffer is known and that
+  /// base buffer is always constructed with the same size.
+  bool hasKnownBaseBufferSize() const { return !baseBufferSize.empty(); }
+
+  /// Returns the constant values for the size of the base buffer if known.
+  const llvm::SmallVector<size_t, 3> &getKnownBaseBufferSize() const {
```
Comment (victor-eds) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h:72]: Same, use `ArrayRef`

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+           "Base buffer size unknown or not constant");
+    return baseBufferSize;
+  }
+
+  /// Returns true if this sub-buffer is always constructed with the same
+  /// offset.
+  bool hasConstantOffset() const { return !subBufOffset.empty(); }
+
+  /// Returns the constant values for the offset of this sub-buffer.
+  const llvm::SmallVector<size_t, 3> &getConstantOffset() const {
```
Comment (victor-eds) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h:83]: Same

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+  bool first = true;
+  BufferInformation info;
+  for (const Definition &def : *mods) {
+    if (first) {
+      info = getInformation(def);
+      first = false;
+    } else {
+      info = info.join(getInformation(def), *aliasAnalysis);
+      if (!info.hasConstantSize() &&
+          info.isSubBuffer() != SubBufferLattice::YES)
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp:190]: Shouldn't we be testing `info.isSubBuffer() == SubBufferLattice::MAYBE`?

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        // information, we can end the processing.
+        return info;
+    }
+  }
+
+  if (pMods) {
+    for (const Definition &def : *pMods) {
+      info = info.join(getInformation(def), *aliasAnalysis);
+      if (!info.hasConstantSize() &&
+          info.isSubBuffer() != SubBufferLattice::YES)
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp:202]: Same

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    return false;
+
+  auto constructor = dyn_cast<sycl::SYCLHostConstructorOp>(def.getOperation());
+  if (!constructor)
+    return false;
+
+  return isa<sycl::BufferType>(constructor.getType().getValue());
+}
+
+BufferInformation SYCLBufferAnalysis::getInformation(const Definition &def) {
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp:224]: Can we rename this to state we're expecting a constructor?

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+      return std::nullopt;
+  }
+
+  bool first = true;
+  BufferInformation info;
+  for (const Definition &def : *mods) {
+    if (first) {
+      info = getInformation(def);
+      first = false;
+    } else {
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp:187]:
```suggestion
  BufferInformation info = getInformation(mods->front());
  for (const Definition &def : mods->drop_front()) {
```
is something like this doable?

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  bool hasConstantSize() const { return !constantSize.empty(); }
+
+  /// Returns the constant values for the size this buffer is constructed with.
+  const llvm::SmallVector<size_t, 3> &getConstantSize() const {
+    return constantSize;
+  }
+
+  /// Returns information about whether this buffer is a sub-buffer of another
+  /// buffer.
+  SubBufferLattice isSubBuffer() const { return subBuf; }
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h:51]: My rationale was that we want to ask "is this a sub-buffer", and the answer is "yes", "no" or "maybe" (encoded in the lattice), so the name seemed appropriate. 

But I can change it if you prefer.

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    return BufferInformation(jointSize, SubBufferLattice::YES, jointBaseBuf,
+                             jointBaseBufSize, jointOffset);
+  }
+
+  // When control flow reaches this point, not both incoming information are
+  // definitely a sub-buffer. If they are both definitely *not* a sub-buffer,
+  // preserve that fact, otherwise be conservative and assume the resulting
+  // information might be a sub-buffer.
+  auto jointSubBuf =
+      (subBuf == other.subBuf) ? SubBufferLattice::NO : SubBufferLattice::MAYBE;
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp]: Good catch, thanks!

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp (hunk 6 above)
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp:227]: This is the private interface, the public interface makes this clear as it's called `getBufferInformationFromConstruction`.

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp (hunk 7 above)
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp:190]: `mods` (and `pMods`) is a `std::set` (defined by the `ReachingDefinitionAnalysis`), so there is no `drop_front` and we would need to use a `for`-loop with explicit iterators to achieve this.

**Review by victor-eds — APPROVED** [review]

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h:51]: My pref would be to change the name but need not block this PR.

**Review by etiotto — APPROVED** [review]

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.h]: Changed the function name to `getSubBuffer` in the latest commit.

**Pull request merged**